syft/pkg/cataloger/common/generic_cataloger.go

```diff
@@ -56,18 +56,44 @@ func (c *GenericCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
 			continue
 		}
 
+		pkgsForRemoval := make(map[artifact.ID]artifact.Identifiable)
 		for _, p := range discoveredPackages {
 			p.FoundBy = c.upstreamCataloger
 			p.Locations.Add(location)
 			p.SetID()
+			// doing it here so all packages have an ID,
+			// IDs are later used to remove relationships
+			if !pkg.IsValid(p) {
+				pkgsForRemoval[p.ID()] = p
+				continue
+			}
+
 			packages = append(packages, *p)
 		}
 
-		relationships = append(relationships, discoveredRelationships...)
+		relationships = removeRelationshipsWithArtifactIDs(pkgsForRemoval, discoveredRelationships)
```

Comment:
Is this logic correct? Since the `append` is removed, this is resetting the `relationships` value on each iteration of the `for` loop, losing prior iterations' relationships.

Reply:
That is a good bug. Thank you for pointing it out. I've fixed it and added a unit test to make it less likely to happen again.